bouncer/src/repo/core/model/bson/repo_node_metadata.cpp

```diff
@@ -29,61 +29,114 @@ RepoNode()
 {
 }
 
-MetadataNode::MetadataNode(RepoBSON bson,
-	const std::unordered_map<std::string, std::pair<std::string, std::vector<uint8_t>>> &binMapping) :
-	RepoNode(bson, binMapping)
+MetadataNode::MetadataNode(RepoBSON bson) :
+	RepoNode(bson)
 {
+	deserialise(bson);
 }
 
 MetadataNode::~MetadataNode()
 {
 }
 
-MetadataNode MetadataNode::cloneAndAddMetadata(
-	const RepoBSON &metadata) const
+void MetadataNode::deserialise(RepoBSON& bson)
 {
-	RepoBSONBuilder metaBuilder, bsonBuilder;
-	metaBuilder.appendElements(metadata);
-	if (hasField(REPO_NODE_LABEL_METADATA))
-	{
-		auto metaBson = getObjectField(REPO_NODE_LABEL_METADATA);
-		if (!metaBson.isEmpty())
-			metaBuilder.appendElementsUnique(metaBson);
+	if (bson.hasField(REPO_NODE_LABEL_METADATA)) {
+		auto arr = bson.getObjectField(REPO_NODE_LABEL_METADATA);
+		auto names = arr.getFieldNames();
+		for (auto& n : names) {
+			auto field = arr.getObjectField(n);
+			auto key = field.getStringField(REPO_NODE_LABEL_META_KEY);
+			if (field.hasField(REPO_NODE_LABEL_META_VALUE))
+			{
+				metadataMap[key] = field.getField(REPO_NODE_LABEL_META_VALUE).repoVariant();
+			}
+			else
```

Comment:
I don't think it's ever possible that we have a key field but no value?

Maybe it's better to throw an error if we find ourselves in this situation

Reply:
I thought the same but it does occur. Looking into it further the `AppendVisitor` filters out empty strings. I don't think this should be the behaviour, so I have removed that check, however existing uploads will still have the missing fields. 
If we add another type to the RepoVariant that can be null we may need to run a migration script in the future.